lib/postgres/postgres-pr/connection.rb

```diff
@@ -343,16 +347,34 @@ def handle_server_error_message(server_error_message)
 
   # tcp://localhost:5432
   # unix:/tmp/.s.PGSQL.5432
-  def establish_connection(uri, proxies)
+  def establish_connection(uri, proxies, ssl = nil)
     u = URI.parse(uri)
     case u.scheme
     when 'tcp'
       @conn = Rex::Socket.create(
-      'PeerHost' => (u.host || DEFAULT_HOST).gsub(/[\[\]]/, ''),  # Strip any brackets off (IPv6)
-      'PeerPort' => (u.port || DEFAULT_PORT),
-      'proto' => 'tcp',
-      'Proxies' => proxies
-    )
+        'PeerHost' => (u.host || DEFAULT_HOST).gsub(/\[|\]/, ''),
+        'PeerPort' => (u.port || DEFAULT_PORT),
+        'proto' => 'tcp',
+        'Proxies' => proxies
+      )
+      if ssl
+        ssl_request_message = SSLRequest.new(80877103)
+        @conn.write(ssl_request_message.dump)
+        response = @conn.read(1)
+        if response == 'S'
+          ssl_context = OpenSSL::SSL::SSLContext.new
+          ssl_socket = OpenSSL::SSL::SSLSocket.new(@conn, ssl_context)
+          # Ensure the underlying TCP socket is closed when the SSL socket is closed
+          # This prevents resource leaks and ensures proper cleanup of the connection
+          ssl_socket.sync_close = true
+          ssl_socket.connect
+          @conn = ssl_socket
+        elsif response == 'N'
+          # Server does not support SSL, continue with plain TCP
```

Comment:
This seems like a vulnerability and an inconsistency to me. If the user specifies that they want to use SSL and can't negotiate it, then they'll have a false sense of security if we maintain and then use the connection. It also seems inconsistent because if you use an HTTP module and set the RPORT to 80 and SSL to true, it doesn't continue on with HTTP request because it can't make HTTPS requests.

I'd suggest raising an exception here.